File: src/main/java/com/rust/util/Shader.java
package com.rust.util;

import com.jogamp.opengl.util.GLBuffers;
import com.sun.jna.platform.win32.OaIdl;
import glm.Glm;
import glm.mat._4.Mat4;
import glm.vec._3.Vec3;
import glm.vec._4.Vec4;

import java.io.*;
import java.nio.ByteBuffer;
import java.nio.FloatBuffer;
import java.nio.IntBuffer;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

import static org.lwjgl.opengl.GL33C.*;

public class Shader {
    public static String VertexShader(String name){
        return "src/main/resources/shaders/vert/" + name + ".vert";
    }
    public static String FragShader(String name){
        return "src/main/resources/shaders/frag/" + name + ".frag";
    }
    // the program ID
    public Integer ID;
    public Shader(String name) throws FileNotFoundException {
        shared_init(VertexShader(name),FragShader(name));
    }
    // constructor reads and builds the shader
    public Shader(String vertexPath, String fragmentPath) throws FileNotFoundException {
        shared_init(vertexPath,fragmentPath);
    }
    private void shared_init(String vertexPath,String fragmentPath) throws FileNotFoundException {
        StringBuffer vertexBuffer = new StringBuffer();
        try{
            FileInputStream vertexStream = new FileInputStream(vertexPath);
            InputStreamReader vertexReader = new InputStreamReader(vertexStream,"UTF-8");
            while (vertexReader.ready()){
                vertexBuffer.append((char)vertexReader.read());
            }
            vertexReader.close();
        }catch (FileNotFoundException e){
            throw new FileNotFoundException("Vertex shader file " + vertexPath + " not found.");
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        Integer vertShader = createShader(GL_VERTEX_SHADER,vertexBuffer.toString());

        StringBuffer fragBuffer = new StringBuffer();
        try{
            FileInputStream fragStream = new FileInputStream(fragmentPath);
            InputStreamReader fragReader = new InputStreamReader(fragStream,"UTF-8");
            while (fragReader.ready()){
                fragBuffer.append((char)fragReader.read());
            }
            fragReader.close();
        }catch (FileNotFoundException e){
            throw new FileNotFoundException("Vertex shader file " + vertexPath + " not found.");
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        Integer fragShader = createShader(GL_FRAGMENT_SHADER,fragBuffer.toString());
        ArrayList<Integer> shaders = new ArrayList<>();
        shaders.add(vertShader);
        shaders.add(fragShader);
        ID = createProgram(shaders);
    }
    // use/activate the shader
    public void use(){
        glUseProgram(ID);
    }
    // utility uniform functions
    public void setBool(String name, Boolean value){
        int intv = value ? 1:0;
        glUniform1i(glGetUniformLocation(ID,name),intv);
    }
    public void setInt(String name, Integer value){
        glUniform1i(glGetUniformLocation(ID,name),value);
    }
    public void setFloat(String name, Float value){
        glUniform1f(glGetUniformLocation(ID,name),value);
    }
    public void setMat4(String name, Mat4 value) {
        glUniformMatrix4fv(glGetUniformLocation(ID,name),false,GLBuffers.newDirectFloatBuffer(value.toFa_()));
    }
    public void setVec3(String name, Vec3 value){
        glUniform3fv(glGetUniformLocation(ID,name),GLBuffers.newDirectFloatBuffer(value.toFa_()));
    }
    public void setVec4(String name, Vec4 value){
        glUniform3fv(glGetUniformLocation(ID,name),GLBuffers.newDirectFloatBuffer(value.toFA_()));
    }
    private int createShader(int shaderType, String shaderFile) {

        int shader = glCreateShader(shaderType);
        glShaderSource(shader, shaderFile);

        glCompileShader(shader);

        IntBuffer status = GLBuffers.newDirectIntBuffer(1);
        glGetShaderiv(shader, GL_COMPILE_STATUS, status);
        if (status.get(0) == GL_FALSE) {

            IntBuffer infoLogLength = GLBuffers.newDirectIntBuffer(1);
            glGetShaderiv(shader, GL_INFO_LOG_LENGTH, infoLogLength);

            ByteBuffer bufferInfoLog = GLBuffers.newDirectByteBuffer(infoLogLength.get(0));
            glGetShaderInfoLog(shader, infoLogLength, bufferInfoLog);
            byte[] bytes = new byte[infoLogLength.get(0)];
            bufferInfoLog.get(bytes);
            String strInfoLog = new String(bytes);
            String strShaderType = "";
            switch (shaderType) {
                case GL_VERTEX_SHADER:
                    strShaderType = "vertex";
                    break;
                case GL_GEOMETRY_SHADER:
                    strShaderType = "geometry";
                    break;
                case GL_FRAGMENT_SHADER:
                    strShaderType = "fragment";
                    break;
            }
            System.err.println("Compiler failure in " + strShaderType + " shader: " + strInfoLog);
        }
        return shader;
    }
    private int createProgram(ArrayList<Integer> shaderList) {

        int program = glCreateProgram();

        shaderList.forEach(shader -> glAttachShader(program, shader));

        glLinkProgram(program);

        IntBuffer status = GLBuffers.newDirectIntBuffer(1);
        glGetProgramiv(program, GL_LINK_STATUS, status);
        if (status.get(0) == GL_FALSE) {

            IntBuffer infoLogLength = GLBuffers.newDirectIntBuffer(1);
            glGetProgramiv(program, GL_INFO_LOG_LENGTH, infoLogLength);

            ByteBuffer bufferInfoLog = GLBuffers.newDirectByteBuffer(infoLogLength.get(0));
            glGetProgramInfoLog(program, infoLogLength, bufferInfoLog);
            byte[] bytes = new byte[infoLogLength.get(0)];
            bufferInfoLog.get(bytes);
            String strInfoLog = new String(bytes);

            System.err.println("Linker failure: " + strInfoLog);
        }
        shaderList.forEach(shader -> glDetachShader(program, shader));
        return program;
    }
}

File: Shader.cpp
#include "Shader.h"
#include <QOpenGLFunctions_3_3_Core>
#include <QOpenGLVersionFunctionsFactory>
Shader::Shader(const char* vertexPath, const char* fragmentPath)
{
    // 1. retrieve the vertex/fragment source code from filePath
    std::string vertexCode;
    std::string fragmentCode;
    std::ifstream vShaderFile;
    std::ifstream fShaderFile;
    // ensure ifstream objects can throw exceptions:
    vShaderFile.exceptions(std::ifstream::failbit | std::ifstream::badbit);
    fShaderFile.exceptions(std::ifstream::failbit | std::ifstream::badbit);
    try
    {
        // open files
        vShaderFile.open(vertexPath);
        fShaderFile.open(fragmentPath);
        std::stringstream vShaderStream, fShaderStream;
        // read file's buffer contents into streams
        vShaderStream << vShaderFile.rdbuf();
        fShaderStream << fShaderFile.rdbuf();
        // close file handlers
        vShaderFile.close();
        fShaderFile.close();
        // convert stream into string
        vertexCode = vShaderStream.str();
        fragmentCode = fShaderStream.str();
    }
    catch (std::ifstream::failure e)
    {
        std::cout << "ERROR::SHADER::FILE_NOT_SUCCESFULLY_READ" << std::endl;
    }
    const char* vShaderCode = vertexCode.c_str();
    const char* fShaderCode = fragmentCode.c_str();

    // 2. compile shaders
    unsigned int vertex, fragment;
    int success;
    char infoLog[512];

    QOpenGLFunctions_3_3_Core* f = QOpenGLVersionFunctionsFactory::get<QOpenGLFunctions_3_3_Core>(QOpenGLContext::currentContext());

    // vertex Shader
    vertex = f->glCreateShader(GL_VERTEX_SHADER);
    f->glShaderSource(vertex, 1, &vShaderCode, NULL);
    f->glCompileShader(vertex);
    // print compile errors if any
    f->glGetShaderiv(vertex, GL_COMPILE_STATUS, &success);
    if (!success)
    {
        f->glGetShaderInfoLog(vertex, 512, NULL, infoLog);
        std::cout << "ERROR::SHADER::VERTEX::COMPILATION_FAILED\n" << infoLog << std::endl;
    };

    // similiar for Fragment Shader
    // vertex Shader
    fragment = f->glCreateShader(GL_FRAGMENT_SHADER);
    f->glShaderSource(fragment, 1, &fShaderCode, NULL);
    f->glCompileShader(fragment);
    // print compile errors if any
    f->glGetShaderiv(fragment, GL_COMPILE_STATUS, &success);
    if (!success)
    {
        f->glGetShaderInfoLog(fragment, 512, NULL, infoLog);
        std::cout << "ERROR::SHADER::VERTEX::COMPILATION_FAILED\n" << infoLog << std::endl;
    };

    // shader Program
    ID = f->glCreateProgram();
    f->glAttachShader(ID, vertex);
    f->glAttachShader(ID, fragment);
    f->glLinkProgram(ID);
    // print linking errors if any
    f->glGetProgramiv(ID, GL_LINK_STATUS, &success);
    if (!success)
    {
        f->glGetProgramInfoLog(ID, 512, NULL, infoLog);
        std::cout << "ERROR::SHADER::PROGRAM::LINKING_FAILED\n" << infoLog << std::endl;
    }

    // delete the shaders as they're linked into our program now and no longer necessary
    f->glDeleteShader(vertex);
    f->glDeleteShader(fragment);
}
void Shader::use()
{
    QOpenGLFunctions_3_3_Core* f = QOpenGLVersionFunctionsFactory::get<QOpenGLFunctions_3_3_Core>(QOpenGLContext::currentContext());
    f->glUseProgram(ID);
}
void Shader::setBool(const std::string& name, bool value) const
{
    QOpenGLFunctions_3_3_Core* f = QOpenGLVersionFunctionsFactory::get<QOpenGLFunctions_3_3_Core>(QOpenGLContext::currentContext());
    f->glUniform1i(f->glGetUniformLocation(ID, name.c_str()), (int)value);
}
void Shader::setInt(const std::string& name, int value) const
{
    QOpenGLFunctions_3_3_Core* f = QOpenGLVersionFunctionsFactory::get<QOpenGLFunctions_3_3_Core>(QOpenGLContext::currentContext());
    f->glUniform1i(f->glGetUniformLocation(ID, name.c_str()), value);
}
void Shader::setFloat(const std::string& name, float value) const
{
    QOpenGLFunctions_3_3_Core* f = QOpenGLVersionFunctionsFactory::get<QOpenGLFunctions_3_3_Core>(QOpenGLContext::currentContext());
    f->glUniform1f(f->glGetUniformLocation(ID, name.c_str()), value);
}

void Shader::setMat4(const std::string& name, glm::mat4 value) const
{
    QOpenGLFunctions_3_3_Core* f = QOpenGLVersionFunctionsFactory::get<QOpenGLFunctions_3_3_Core>(QOpenGLContext::currentContext());
    f->glUniformMatrix4fv(f->glGetUniformLocation(ID, name.c_str()), 1, GL_FALSE, glm::value_ptr(value));
}

void Shader::setVec(const std::string& name,int size, float* value) const
{
    QOpenGLFunctions_3_3_Core* f = QOpenGLVersionFunctionsFactory::get<QOpenGLFunctions_3_3_Core>(QOpenGLContext::currentContext());
    f->glUniform3fv(f->glGetUniformLocation(ID, name.c_str()), size, value);
}
void Shader::setVec3(const std::string& name, glm::vec3 value) const
{
    QOpenGLFunctions_3_3_Core* f = QOpenGLVersionFunctionsFactory::get<QOpenGLFunctions_3_3_Core>(QOpenGLContext::currentContext());
    f->glUniform3fv(f->glGetUniformLocation(ID, name.c_str()),1,&value[0]);
}